runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java

```diff
@@ -60,13 +55,36 @@
   private final DataTransferFactory channelFactory;
   private final MetricCollector metricCollector;
 
-  /**
-   * Map of task IDs in this task group to their readers/writers.
-   */
-  private final Map<String, List<InputReader>> physicalTaskIdToInputReaderMap;
-  private final Map<String, List<OutputWriter>> physicalTaskIdToOutputWriterMap;
-
-  private boolean isExecutionRequested;
+  // Map of task ID to its intra-TaskGroup data pipe.
+  private final Map<Task, List<LocalPipe>> taskToInputPipesMap;
+  private final Map<Task, LocalPipe> taskToOutputPipeMap;  // one and only one Pipe per task
+  // Readers/writers that deals with inter-TaskGroup data.
+  private final List<InputReader> inputReaders;
+  private final Map<Task, List<InputReader>> taskToInputReadersMap;
+  private final Map<Task, List<InputReader>> taskToSideInputReadersMap;
+  private final Map<Task, List<OutputWriter>> taskToOutputWritersMap;
+  private final Map<InputReader, List<Task>> inputReaderToTasksMap;
+  private final Map<String, Iterator> idToSrcIteratorMap;
+  private final Map<String, List<Task>> srcIteratorIdToTasksMap;
+  private final Map<String, List<Task>> iteratorIdToTasksMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> iteratorQueue;
+  private volatile Map<String, List<Task>> pipeIdToDstTasksMap;
```

Comment:
Can you explain in the comment why this is marked as `volatile`? My understanding is that only a single thread instantiates and executes a `TaskGroupExecutor`. Let me know if I'm missing something.